tools/depcheck/pkg/cmd/trace.go

```diff
@@ -0,0 +1,140 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/encoding/dot"
+
+	depgraph "github.com/openshift/origin/tools/depcheck/pkg/graph"
+)
+
+var traceImportsExample = `# create a dependency graph
+%[1]s trace --entry=pkg/foo --root=pkg/bar
+
+# create a dependency graph and output in "dot" format
+%[1]s trace --entry=pkg/foo --output=dot
+`
+
+type TraceOptions struct {
+	GraphOptions *depgraph.GraphOptions
+
+	OutputFormat string
+
+	Out    io.Writer
+	ErrOut io.Writer
+}
+
+type TraceFlags struct {
+	GraphFlags *depgraph.GraphFlags
+
+	OutputFormat string
+}
+
+func (o *TraceFlags) ToOptions(out, errout io.Writer) (*TraceOptions, error) {
+	graphOpts, err := o.GraphFlags.ToOptions(out, errout)
+	if err != nil {
+		return nil, err
+	}
+
+	return &TraceOptions{
+		GraphOptions: graphOpts,
+		OutputFormat: o.OutputFormat,
+
+		Out:    out,
+		ErrOut: errout,
+	}, nil
+}
+
+func NewCmdTraceImports(parent string, out, errout io.Writer) *cobra.Command {
+	traceFlags := &TraceFlags{
+		GraphFlags:   &depgraph.GraphFlags{},
+		OutputFormat: "dot",
+	}
+
+	cmd := &cobra.Command{
+		Use:     "trace --entry=pkg/foo",
+		Short:   "Creates a dependency graph from a set of given entrypoint packages",
+		Long:    "Creates a dependency graph from a set of given entrypoint packages",
+		Example: fmt.Sprintf(traceImportsExample, parent),
+		RunE: func(c *cobra.Command, args []string) error {
+			opts, err := traceFlags.ToOptions(out, errout)
+			if err != nil {
+				return err
+			}
+
+			if err := opts.Complete(); err != nil {
+				return err
+			}
+			if err := opts.Validate(); err != nil {
+				return err
+			}
+			if err := opts.Run(); err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	traceFlags.GraphFlags.AddFlags(cmd)
+	cmd.Flags().StringVarP(&traceFlags.OutputFormat, "output", "o", traceFlags.OutputFormat, "output generated dependency graph in specified format. One of: dot.")
+	return cmd
+}
+
+func (o *TraceOptions) Complete() error {
+	return o.GraphOptions.Complete()
+}
+
+func (o *TraceOptions) Validate() error {
+	if err := o.GraphOptions.Validate(); err != nil {
+		return err
+	}
+
+	if len(o.OutputFormat) > 0 && o.OutputFormat != "dot" {
+		return fmt.Errorf("invalid output format provided: %s", o.OutputFormat)
+	}
+
+	return nil
+}
+
+// Run execs `go list` on all package entrypoints specified through --root.
+// Each package's ImportPath and non-transitive (immediate) imports are
+// filtered and aggregated. A package is filtered based on whether its ImportPath
+// matches an accepted pattern defined in the set of validPackages.
+// Each aggregated package becomes a node in a generated dependency graph.
+// An edge is created between a package and each of its Imports.
+func (o *TraceOptions) Run() error {
+	g, err := o.GraphOptions.BuildGraph()
+	if err != nil {
+		return err
+	}
+
+	if len(o.GraphOptions.Filters) > 0 {
```

Comment:
This filtering should have happened before you gave me back the graph. Move to BuildGraph.